sys/linux/usbip.txt

```diff
@@ -0,0 +1,47 @@
+# Copyright 2020 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <drivers/usb/usbip/usbip_common.h>
+
+resource usbip_server_fd[fd]
+
+syz_usbip_server_init(usb3 bool32, speed flags[usbip_device_speed]) usbip_server_fd
+write$usbip(fd usbip_server_fd, buffer ptr[in, usbip_header], len bytesize[buffer])
+
+type usbip_header_basic[PACKET_TYPE] {
+	command		const[PACKET_TYPE, int32be]
+	seqnum		int32be
+	devid		int32be
+	direction	flags[direction_flags, int32be]
+	ep		int32be
+} [packed]
+
+usbip_header_ret_submit {
+	status			int32be
+	actual_length		int32be
+	start_frame		int32be
+	number_of_packets	int32be[0:USBIP_MAX_ISO_PACKETS]
+	error_count		int32be
+} [packed]
+
+usbip_header_ret_unlink {
+	status	int32be
+} [packed]
+
+usbip_packet_ret_submit {
+	base_submit	usbip_header_basic[USBIP_RET_SUBMIT]
+	ret_submit	usbip_header_ret_submit
+} [packed]
+
+usbip_packet_ret_unlink {
+	base_unlink	usbip_header_basic[USBIP_RET_UNLINK]
+	ret_unlink	usbip_header_ret_unlink
+} [packed]
+
+usbip_header [
```

Comment:
Running tools/syz-check, I've got the following warnings. All look true.
```
no-such-struct: usbip_packet_ret_submit
no-such-struct: usbip_packet_ret_unlink
bad-struct-size: usbip_header: syz=40 kernel=48
```

Reply:
The size mismatch won't allow to get any coverage I think because kernel checks for exact `usbip_header` size on receive.

Reply:
I think this is what we should do here:

```
define USBIP_HEADER_SIZE sizeof(struct usbip_header)

usbip_header [
...
] [packed, size[USBIP_HEADER_SIZE]]
```

Reply:
Ah, it's actually more tricky. We also have data following the header for `USBIP_RET_SUBMIT`, see [this](https://elixir.bootlin.com/linux/latest/source/drivers/usb/usbip/vhci_rx.c#L55). The amount of data also depends on the fields values in the basic header.